code/modules/vehicles/mecha/combat/justice.dm

```diff
@@ -0,0 +1,455 @@
+			return
+	return ..()
+
+/datum/action/vehicle/sealed/mecha/invisibility
+	name = "Invisibility"
+	button_icon_state = "mech_stealth_off"
+	/// Is invisibility activated.
+	var/on = FALSE
+	/// Recharge check.
+	var/charge = TRUE
+	/// Block movement if we start aoe attack from invisibility.
+	var/start_attack = FALSE
+	/// Aoe pre attack sound.
+	var/stealth_pre_attack_sound = 'sound/mecha/mech_stealth_pre_attack.ogg'
+	/// Aoe attack sound.
+	var/stealth_attack_sound = 'sound/mecha/mech_stealth_attack.ogg'
+
+/datum/action/vehicle/sealed/mecha/invisibility/Trigger(trigger_flags)
+	. = ..()
+	if(chassis.weapons_safety)
+		owner.balloon_alert(owner, "safty is on!")
+		return
+	if(!charge)
+		owner.balloon_alert(owner, "on recharge!")
+		return
+	new /obj/effect/temp_visual/mech_sparks(get_turf(chassis))
+	on = !on
+	playsound(chassis, 'sound/mecha/mech_stealth_effect.ogg' , 75, FALSE)
+	if(on)
+		for(var/mob/living/occupant in chassis.occupants)
+			var/datum/action/vehicle/sealed/mecha/charge_attack/charge_action = LAZYACCESSASSOC(chassis.occupant_actions, occupant, /datum/action/vehicle/sealed/mecha/charge_attack)
+			if(charge_action?.on)
+				charge_action.on = FALSE
+		animate(chassis, alpha = 0, time = 0.5 SECONDS)
+		button_icon_state = "mech_stealth_on"
+		addtimer(CALLBACK(src, PROC_REF(end_stealth)), 20 SECONDS)
+		RegisterSignal(chassis, COMSIG_MECHA_MELEE_CLICK, PROC_REF(stealth_attack_aoe))
+	else
+		addtimer(CALLBACK(src, PROC_REF(charge)), 5 SECONDS)
+		animate(chassis, alpha = 255, time = 0.5 SECONDS)
+		button_icon_state = "mech_stealth_off"
+		UnregisterSignal(chassis, COMSIG_MECHA_MELEE_CLICK)
+	build_all_button_icons()
+
+/**
+ * ## end_stealth
+ *
+ * Called when mech runs out of invisibility time.
+ */
+/datum/action/vehicle/sealed/mecha/invisibility/proc/end_stealth()
+	if(!on)
+		return
+	owner.balloon_alert(owner, "invisability is over")
+	Trigger()
+
+/**  Proc makes an AOE attack after 1 SECOND.
+ * Called by the mech pilot when he is in stealth mode and wants to attack.
+ * During this, mech cannot move.
+*/
+/datum/action/vehicle/sealed/mecha/invisibility/proc/stealth_attack_aoe(datum/source, mob/living/pilot, atom/target, on_cooldown, is_adjacent)
+	SIGNAL_HANDLER
+
+	if(!charge)
+		return FALSE
+	if(chassis.alpha != 0)
+		UnregisterSignal(chassis, COMSIG_MECHA_MELEE_CLICK)
+		return FALSE
+	UnregisterSignal(chassis, COMSIG_MECHA_MELEE_CLICK)
+	new /obj/effect/temp_visual/mech_attack_aoe_charge(get_turf(chassis))
+	start_attack = TRUE
+	playsound(chassis, stealth_pre_attack_sound, 75, FALSE)
+	addtimer(CALLBACK(src, PROC_REF(attack_in_aoe), pilot), 1 SECONDS)
+	return TRUE
+
+/**
+ * ## attack_in_aoe
+ *
+ * Brings mech out of invisibility.
+ * Deal everyone in range 3x3 35 damage and 25 chanse to cut off limb.
+ * Arguments:
+ * * pilot - occupant inside mech.
+ */
+/datum/action/vehicle/sealed/mecha/invisibility/proc/attack_in_aoe(mob/living/pilot)
+	Trigger()
```

Comment:
If this is meant to make the mech visible, use a proc which makes the mech visible instead and then start the cooldown.